### ISSUE TITLE
enhancements

### ISSUE THREAD

**Issue by janrito** [jgm/pandoc#613]
Let me start by saying that Pandoc's markdown flavour is great, it is flexible enough to have added obviously beneficial extensions to the original language, like citations, blocked code and examples. It also has been conservative enough not to allow feature creep and let it become unreadable and ugly. 

Markdown has become not only a fantastic markup in which read and archive documents, but more importantly, a markup which is unobtrusive and confortable to write original material. Pandoc's Markdown has made it more versatile and complete. 

However, there are some areas where the language is lacking. I would like to add this issue to open a discussion. I think I've identified some issues, and rather than just offer hack solutions, I would like to open this for suggestions. Maybe I wont make the best case for how to fix them, but I think I've identified some problems that others might share.
# Object labels

The way that labeling works is inconsistent, and their function varies per object.
- Headers (automatically)
  
  Headers are given an automatic reference based on a stripped down and cleaned version of the text, and may be referenced later in the text via a standard link to an id:
  
  ``` markdown
  See the [Introduction](#introduction).
  ```
- Links:
  
  ``` markdown
    Here is [my link][FOO]
  
    [Foo]: /bar/baz
  ```
- Images: 
  
  ``` markdown
    ![movie reel]
  
    [movie reel]: movie.gif
  ```
  
  Images are sort of labelled, the url can be added as images, but they cannot be referenced from anywhere else in the document.
- Code blocks:
  
  <pre>

``` #mycode .haskell .numberLines startFrom="100"
qsort []     = []
qsort (x:xs) = qsort (filter (&lt; x) xs) ++ [x] ++
               qsort (filter (&gt;= x) xs)
```

  </pre>

  Code blocks can be referenced elsewhere in the document in the same way as headers:

``` markdown
  See [example](#mycode).
```
- example list items:
  
  ``` markdown
  (@good)  This is a good example.
  
  As (@good) illustrates, ...
  ```

So, to summarise, labels are attached to objects with inconsistent syntax  `[#label]`, `![label]`, `[][label]`, `{#label}`, `(@label), and sometimes are implicit. They are sometimes referenceable, and sometimes not.

On the other hand, there are some objects that would clearly benefit from having lables, but don't.
- Tables:
  
  ``` markdown
  -------------------------------------------------------------
  Centered   Default           Right Left
  Header    Aligned         Aligned Aligned
  ----------- ------- --------------- -------------------------
     First    row                12.0 Example of a row that
                                      spans multiple lines.
  
    Second    row                 5.0 Here's another one. Note
                                      the blank line between
                                      rows.
  -------------------------------------------------------------
  
  Table: Here's the caption. It, too, may span
  multiple lines.
  ```
  
  `Table` in this context only tells the parser that the next sentence is a caption for the table. In fact, it is unnecessary, a single `:` can be used. However this space could be used to add a label to the table and allow it to be referenced elsewhere in the document.
- Math:
  
  ``` markdown
  $f(x) = ax^2 + bx + c$
  ```

Some of these have already been included in other versions of markdown, like multimarkdown. An image label is attached to an image instead of the url in it's deffinition:

``` markdown
![This is a *caption*][fig2]

I like cats [](#fig2).
```

Tables, also have a labeling syntax in multimarkdown:

``` markdown
|             |          Grouping           ||
First Header  | Second Header | Third Header |
 ------------ | :-----------: | -----------: |
Content       |          *Long Cell*        ||
Content       |   **Cell**    |         Cell |

New section   |     More      |         Data |
And more      |            And more         ||
[Prototype_table]: with caption

Cells are nice [](#Prototype_table).
```
# Block objects
- Figures:
  Images are wrapped in a figure element in html and latex when they are in their own paragraph. Often, however, figures are made up of a collection of images. A block of contiguous images, could be treated as a single figure and labeled individually.
  
  For example,
  
  ``` markdown
      ![I love cats][#cat_love]
      ![The cat ate my shoelaces][#cat_hate]
      [fig1]: My relationship with cats is complicated
  
      In this example [](#fig1)
  
      [cat_love]: /url/to/cat1.jpg 
      [cat_hate]: /url/to/cat2.jpg 
  ```
- Math:
  
  Math objects are currently treated as inline elements, but just as images, sometimes a they might need to be treated as a block. Allowing for labelling of a block math object can be translated into latex as an equation.  
  
  ``` markdown
  $
  f(x) = ax^2 + bx + c
  $
  [equation1]
  
  Math is useful [](#equation1)
  ```

**TLDR;** Pandoc is awesome, but everything sucks. I mean, I love you but You could so much be better.
